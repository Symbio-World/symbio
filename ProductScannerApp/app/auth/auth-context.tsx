import React, { useEffect, useState, useContext } from 'react'

export type User = {
  uid: string
}

type AuthContext = {
  user?: User
}

const authContext = React.createContext<AuthContext>({})

type Props = {
  children: React.ReactNode
}

type Deps = {
  signInAnonymously: () => Promise<{ user: User }>
}

type CreateAuthProvider = (deps: Deps) => React.FC<Props>

export const createAuthProvider: CreateAuthProvider = ({
  signInAnonymously
}) => ({ children }: Props) => {
  const [user, setUser] = useState<User>()

  const authenticate = async () => {
    console.log('signInAnonymously', signInAnonymously)
    const { user } = await signInAnonymously()
    console.log('user', user)
    setUser(user)
  }

  useEffect(() => {
    authenticate()
  }, [])

  console.log('user', user)
  return (
    <authContext.Provider value={{ user }}>{children}</authContext.Provider>
  )
}

export const useAuth = () => useContext(authContext)
