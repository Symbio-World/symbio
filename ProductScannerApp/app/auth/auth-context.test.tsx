import 'react-native'
import React from 'react'
import { Text } from 'react-native'
import { create, act, ReactTestRenderer } from 'react-test-renderer'

import { createAuthProvider, useAuth } from './auth-context'

describe('AuthContext', () => {
  it('renders correctly', () => {
    const signInAnonymously = jest.fn()
    const AuthProvider = createAuthProvider({ signInAnonymously })
    create(<AuthProvider>Test</AuthProvider>)
    expect(signInAnonymously).toHaveBeenCalled
  })

  it('can be consumed', async () => {
    const promise = Promise.resolve({ user: { uid: 'uid' } })
    const signInAnonymously = jest.fn<any, any>(() => promise)
    const AuthProvider = createAuthProvider({ signInAnonymously })
    const Child = () => {
      const { user } = useAuth()

      return <Text>{user?.uid}</Text>
    }
    let root: ReactTestRenderer
    act(() => {
      root = create(<AuthProvider><Child /></AuthProvider>)
    })

    expect(root!.toJSON()).toMatchSnapshot();

    await act(async () => {
      await promise
    })

    expect(root!.toJSON()).toMatchSnapshot();
  })
})
