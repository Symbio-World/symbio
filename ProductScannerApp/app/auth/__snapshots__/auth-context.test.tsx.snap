// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthContext can be consumed 1`] = `<Text />`;

exports[`AuthContext can be consumed 2`] = `
<Text>
  uid
</Text>
`;
